client/components/PlanningTemplatesModal/TemplatesListView.tsx

```diff
@@ -0,0 +1,75 @@
+import {ICalendar, IEventTemplate} from 'interfaces';
+import React from 'react';
+import {gettext} from 'superdesk-core/scripts/core/utils';
+import {Heading, BoxedList, BoxedListItem} from 'superdesk-ui-framework/react';
+
+type ITemplatesListViewProps = {
+    closeModal: () => void;
+    eventTemplates: Array<IEventTemplate>;
+    calendars: Array<ICalendar>;
+    activeCalendarFilter?: string;
+    searchQuery: string;
+    createEventFromTemplate: (template: IEventTemplate) => void;
+}
+
+export const TemplatesListView: React.FC<ITemplatesListViewProps> = ({
+    eventTemplates,
+    calendars,
+    closeModal,
+    activeCalendarFilter,
+    searchQuery,
+    createEventFromTemplate,
+}: ITemplatesListViewProps) => {
+    /**
+     * Groups the templates by calendar,
+     * filters the templates that match the current search query,
+     * if a calendar is selected filters the groups that match the calendar are filtered,
+     * if not the groups that that don't have any templates are filtered out.
+     */
+    const filteredTemplates = calendars
```

Comment:
![image](https://github.com/superdesk/superdesk-planning/assets/2076953/b6e1386a-8eb0-40d4-8447-4b2adb564ba7)


This is very inefficient. It will probably not cause significant performance issues, but try getting used to writing code in an efficient way when it's not any harder like here. One thing to keep in mind is minimize iterations. You will not remove iterations entirely - but when you iterate, make sure you are iterating over less and less items in every iteration. Another rule is to avoid coupling which you did well here.

here's an example how you can do it more efficiently

```
const templatesMatchingSearchQuery = eventTemplates.filter((template) => template.template_name.includes(searchQuery));

const templatesByCalendar = {}; // only those matching search query

for (const template of templatesMatchingSearchQuery) {
    for (const calendar of template.data.calendars ?? []) {
        if (templatesByCalendar[calendar.qcode] == null) {
            templatesByCalendar[calendar.qcode] = [];
        }

        templatesByCalendar[calendar.qcode].push(template);
    }
}

const filteredCalendars = activeCalendarFilter == null
    ? props.calendars
    : [props.calendars.find(({qcode}) => qcode === activeCalendarFilter)];

const filteredCalendarsWithTemplates = filteredCalendars.map((calendar) => ({
    calendar: calendar,
    templates: templatesByCalendar[calendar.qcode],
}));
 ```
 
It's also not only about of number of iterations but also how performance intensive is each iteration. For example, I could have first grouped templates by calendars and only then performed filtering by template name on matching templates. If one calendar is selected in UI that text comparison would be done only on templates in one calendar.